_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+- Issuer - In this RFC this refers to the issuer field in a OIDC metadata
+  sheet. The issuer is an authority for validating JWT tokens used in OIDC.
+
+## Motivation
+
+Some claim a good feature of secure infrastructure is access control. Currently,
+our only solution is network isolation. Being able to restrict access to varying
+degrees is also a product goal to satisfy business needs.
+
+Tying the user management of wharf to central repositories such as
+organisation wide IDPs is good for a number of reasons.
+- Centrally managed on/off boarding:
+   - Leads to less administration
+   - Roll based security inheritance allows automatic updates of user access
+     levels
+- Piggybacking on existing password and user recovery routines results in a
+  DRY style system architecture
+- Near perfectly mirrored compliance with top level organisational
+  security levels
+
+## Explanation
+
+By using the OIDC standard an authentication & authorisation system will be
+built in a minimalist fashion that offloads most of the work involved in
+administration to other pre-existing platforms.
+
+![Connectivity visualised](../assets/auth-map-v2.PNG)
+
+#### The ID Token
+A OIDC type ID-token can be used for authentication. The custom claims
+(requires an IDP with support for RBAC) specified within the ID-token can
+then be used to specify access levels. To keep the scope manageable the task
+of dividing up and refining access levels and permission will be left to a
+later date. For this RFC all kinds of read permissions won't require valid
+ID-tokens, but all write type requests will require valid ID-tokens that
+have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[ID Token](https://auth0.com/docs/tokens/id-tokens/id-token-structure).
+The ID token is signed by the source issuer. This signature is then used to
+validity and authenticity of the token. See
+[validating an OIDC ID-Token](https://connect2id.com/blog/how-to-validate-an-openid-connect-id-token).
+
+#### Sessions
+The wharf-api is currently stateless and does not use sessions. It is
+possible to either include the ID token in every request to the server or to
+start using sessions to store the ID token between requests.
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
```

Comment:
The problem is the potential preformance risk of having an excessive number of requests fetching the IDP issuer keys.
No caching of the public keys for the IDP does not relate to sessions. Without caching the issuer public keys for the IDP a request to fetch would have to be made for every authorization request (or in the stateless session variety - every request). 
If we kept sessions then we would only have to authenticate once per session. This would almost entierly rid us of the imminent preformance problem.

Reply:
Perhaps we shold a brainstomring meeting on this one.

Reply:
I have done som poking around the problem on this one and this seems to be a very minor problem which is probably best solved later and left as simple as possible now.